### ISSUE TITLE
Using yarp's internal logging for yarp devices

### ISSUE THREAD

**Issue by David-Estevez** [roboticslab-uc3m/best-practices#11]
Hi!

I have just discovered that yarp has a [logging system for the log messages](http://www.yarp.it/Log_8h_source.html) (info, debug, warning, error...) of their yarp devices. We are using a [similar system](https://github.com/roboticslab-uc3m/color-debug) for our devices, based on `color-debug`, but I guess their logging system has better integration with yarp devices and other yarp mechanisms. Messages would also look more homogeneous if sent to stdout with the yarp logger instead of our own system.

The question is... should we use yarp's internal logging for our devices? or should we stick to the current `color-debug` system?

**Comment by PeterBowman** [comment]
I recall having run across some YARP utilities that understand and handle these logging streams, perhaps `yarpdatadumper` or `yarpdataplayer`. We can also mention the `yarplogger` program, but [its documentation is scarce](http://www.yarp.it/yarplogger.html).

A few cons against migrating to YARP's logger:

* No colours (obviously).
* I don't see any option for enabling or disabling specific logging macros (you can control this via CMake with color-debug).
* ...?

**Comment by jgvictores** [comment]
Behold a screenshot of `yarplogger`.

![yarplogger](https://user-images.githubusercontent.com/4448588/28534641-a5764a58-70a1-11e7-89e0-2b30a00a6e14.png)

**Comment by David-Estevez** [comment]
It reminds me of Android's Logcat, which is extremely useful for debugging on Android:

![image](https://user-images.githubusercontent.com/1451819/28569084-9b72c1fc-7138-11e7-9dce-657f55e217fd.png)

**Comment by PeterBowman** [comment]
> No colors (obviously).

False, just set the `YARP_COLORED_OUTPUT` environment variable. See INFO and DEBUG logs below:

![Screenshot from 2019-08-01 21-37-34](https://user-images.githubusercontent.com/9977198/62322163-b5d85280-b4a4-11e9-89a0-d1908655b5c6.png)

**Comment by PeterBowman** [comment]
A few more words about YARP log streams:

* six levels: trace (if `YARP_TRACE_ENABLE=1`), info, warning, debug, error, fatal (calls `std::exit`)
* supports colors (if `YARP_COLORED_OUTPUT=1`)
* verbose output (if `YARP_VERBOSE_OUTPUT=1`) prints full paths and function signatures along with the line number
* broadcasts logs via YARP port (if `YARP_FORWARD_LOG_ENABLE=1` or `yarplogger` was started)
* you can't disable specific log levels on the console
* nice `yarplogger` support, didn't try `yarplogger-console` yet
* recommended way of using `yarplogger` is to run apps/devices in a `yarprun` node (add `--log`); thus, every single output is recorded, whereas the `YARP_FORWARD_LOG_ENABLE=1` way is unable to record non-YARP-logger lines
* supports two signatures per log level:
  * `yDebug(...)`: much like `printf(...)`, inserts a newline at the end
  * `yDebug() << "value:" << val;`: stream-like format, you can store this in a variable and append more stuff later within the same line of text (remember `CD_DEBUG_NO_HEADER`?)
* `yarplogger` allows you to access logs for specific ports, show them on GUI with timestamps and nice colors (you can disable unwanted log levels), save specific logs to file or the entire log session

I'm a bit concerned about performance and the lack of Windows support (we can take advantage of previous experience on color-debug and submit a patch, though; see https://github.com/roboticslab-uc3m/color-debug/issues/5), but this is really cool. Shall we take this seriously and replace color-debug usage in our repos? Attaching screenshot for science, I made color-debug macros call YARP logger functions for this test:

![Screenshot from 2019-08-01 23-06-25](https://user-images.githubusercontent.com/9977198/62327487-0eade800-b4b1-11e9-8e6b-54bb75f8e593.png)

Relevant sources:

* [yarp/os/Log.h](https://github.com/robotology/yarp/blob/b0064411073802eb826f9523c0f014dda7e89dac/src/libYARP_OS/include/yarp/os/Log.h) (include this if using `yDebug(...)` signature)
* [yarp/os/LogStream.h](https://github.com/robotology/yarp/blob/b0064411073802eb826f9523c0f014dda7e89dac/src/libYARP_OS/include/yarp/os/LogStream.h) (include this if using `yDebug() << ...` signature)
* [Log.cpp](https://github.com/robotology/yarp/blob/b0064411073802eb826f9523c0f014dda7e89dac/src/libYARP_OS/src/Log.cpp)

**Comment by PeterBowman** [comment]
YARP's logging system is undergoing a major overhaul. New features include fine-grained component-based tagging, and a lot more information accompanying each log event. See https://github.com/robotology/yarp/pull/2256 and linked dependent tickets. Also, https://github.com/robotology/yarp/pull/2258 contains a nice screenshot of new *yarplogger* in action.

> I'm a bit concerned about performance and the lack of Windows support (we can take advantage of previous experience on color-debug and submit a patch, though; see [roboticslab-uc3m/color-debug#5](https://github.com/roboticslab-uc3m/color-debug/issues/5)), but this is really cool.

Implemented in YARP 3.3.x, see https://github.com/robotology/yarp/pull/2076.

**Comment by PeterBowman** [comment]
> YARP's logging system is undergoing a major overhaul.

This is nicely documented in http://www.yarp.it/yarp_logging.html ([source](https://github.com/robotology/yarp/blob/master/doc/yarp_logging.md)).

**Comment by PeterBowman** [comment]
Currently running YARP 3.4 on TEO (https://github.com/roboticslab-uc3m/teo-hardware-issues/issues/58) with the new logging system and fancy colors enabled (https://github.com/roboticslab-uc3m/teo-developer-manual/commit/89c44dcfdc93e13dc49348774be1c48169151927).

**Comment by PeterBowman** [comment]
In view of the forthcoming replacement of color-debug with YARP logging goodies, I'm transferring this issue to a more horizontally focused location.
